auth/authzn.proto

```diff
@@ -0,0 +1,179 @@
+// gRPC connection) with a definition of an authentication and authorization
+// policy that should be used for the respective (authn/authz) of the relevant
+// service. 
+//
+// It is expected that when authorization and authentication details are
+// supplied via the gNOI Authzn service, this config cannot be manipulated
+// through other interfaces, such that a restriction on the gNOI.authzn RPC
+// itself restricts the set of users that are able to manipulate the auth[zn]
+// configuration of the end device.
+//
+// In order to ensure that the auth[zn] configuration supplied in the service
+// cannot be subverted by changing the configuration of the device via gNMI,
+// the following leaves with the OpenConfig system model should be ignored:
+// 
+//  - /system/aaa/authentication/config/authentication-method
+//  - /system/aaa/authorization/config/authorization-method
+//  - /system/aaa/authentication/admin-user/...
+//  - /system/aaa/authentication/users/...
+//
+// The applied state (i.e., the 'state' version of the paths above) should
+// reflect a new authentication-method and authorization-method of GNOI_AUTHZN
+// which is to be added as a new identity with the base of
+// openconfig-aaa-types:AAA_METHOD_TYPE (http://ops.openconfig.net/branches/models/master/docs/openconfig-system.html#ident-aaa_method_type)
+//
+// The motivation for this service is provided in the accompanying
+// authn-authz-intro.md document.
+service Authzn {
+    // Set allows a user to specify an authentication and authorization
+    // policy to the network element. The policy, if valid, replaces the
+    // existing policy.
+    //
+    // Canonical error codes are to be used as responses. See
+    // https://grpc.io/docs/guides/error/ for more details.
+    rpc Set(SetRequest) returns (SetResponse) {}
+}
+
+message SetRequest {
+    // A list of policies that are to be applied on the network element.
+    // Multiple policies per service MAY be specified to allow for multiple
+    // users, or specific user restrictions to be described.
+    repeated Policy policy = 1;
+}
+
+// Policy specifies an authentication/authorization policy for an individual
+// service that is instantiated on the device. Some services may only have an
+// authentication policy - where they do not support more granular authorization,
+// whereas others may have both specified.
+message Policy {
+    oneof protocol {
+        // console provides the configuration for authentication of the local
+        // system login.
+        Console console = 1;
+        // ssh provides the configuration for authentication of logins to the
+        // system via SSH.
+        SSH ssh = 2;
+        // gnmi provides a local authentication mechanism for all RPCs within
+        // the gNMI service, along with an authorization policy which allows
+        // users to be restricted to:
+        //  - specific RPCs (e.g., Get, Set, Subscribe, Capabilities)
```

Comment:
Restricting users to specific RPCs overlaps what gRPC will provide.  How do you see the two schemes work and be managed together?

Reply:
My assumption here was that:

* gRPC would provide us the means to implement this authentication and authorization
* this service would provide the means to provide the policy that gRPC enforces

Is there something that explains what the proposed gRPC mechanisms are -- for example, is there already a means by which gRPC is proposing to have the per-RPC authorisation specified?

Reply:
Actually the gRPC authorization framework will include both policy specification and enforcement.  I will share the proposal with you separately.

Sorry for the delay in replying.  I must have have the right setup for this Github project that I am not getting email for your replies.